[FSDP][optim_state_dict][7/N] Make FSDP support NamedOptimizer

fegin opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #91234
* __->__ #91160
* #90798

**What does this PR do?**
This PR refactors FSDP optimizer state_dict APIs to accept `NamedOptimizer` as the input optimizer. The key difference is that the state_dict returned by `NamedOptimizer` is already keyed as FQN. This PR majorly changes the internal mapping to allows the optimizer state_dict to be keyed as FQN.

fegin committed:
[FSDP][optim_state_dict][7/N] Make FSDP support NamedOptimizer

[ghstack-poisoned]

awgu commented on torch/distributed/fsdp/_optim_utils.py:
```diff
@@ -1396,16 +1448,19 @@ def _gather_orig_param_state(
-                    assert torch.equal(curr_value, value)
-                else:
-                    orig_state[state_name] = value
+                    curr_value.append(value.to(device))
```
This is not in scope for this PR, but I wonder if this will break CPU offloading, which runs the optimizer step on CPU.

awgu reviewed (commented):
I wonder if it worthwhile to document what the assumptions are for how the parameter IDs are chosen for a given optimizer and where those assumptions are needed (e.g. we assume that the optimizer enumerates the IDs in parameter group order, both within and across parameter groups).

fduwjj commented on torch/distributed/fsdp/_optim_utils.py:
```diff
@@ -1171,23 +1214,26 @@ def _map_param_id_to_optim_keys(
 def _unflatten_process_groups(
     state_dict: Dict[str, Any],
-    param_id_to_param: List[nn.Parameter],
+    param_key_to_param: List[nn.Parameter],
```
Do we want to change the type of this variable?

fduwjj reviewed (approved):
Overall lgtm.

fduwjj commented:
@awgu that's a good point. From the result I played with optimizer. Looks you cannot specify both all parameters and param_groups at the same time. And the order is either the order of all parameters or param_groups.

fegin committed:
Update on "[FSDP][optim_state_dict][7/N] Make FSDP support NamedOptimizer"




[ghstack-poisoned]

fegin committed:
Update on "[FSDP][optim_state_dict][7/N] Make FSDP support NamedOptimizer"


**What does this PR do?**
This PR refactors FSDP optimizer state_dict APIs to accept `NamedOptimizer` as the input optimizer. The key difference is that the state_dict returned by `NamedOptimizer` is already keyed as FQN. This PR majorly changes the internal mapping to allows the optimizer state_dict to be keyed as FQN.


[ghstack-poisoned]

rohan-varma reviewed (approved):
stamp to unblock

fegin commented on test/distributed/fsdp/test_fsdp_optim_state.py:
```diff
@@ -1439,6 +1440,70 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
+        self._check_same_param_groups(
+            state_dicts[0], state_dicts[1], check_same_param_keys=False
+        )
+        self._check_same_state(
```
No. Only `optim1` steps 5 times so there will be difference. However, we then load the state_dict back so that we should get the same state. This logic is to test the correctness of loading.

fegin commented on test/distributed/fsdp/test_fsdp_optim_state.py:
```diff
@@ -1439,6 +1440,70 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
+            state_dicts[0], state_dicts[1], check_same_param_keys=False
+        )
+        self._check_same_state(
+            state_dicts[0], state_dicts[1], check_same_param_keys=True
```
The  `state["param_groups"]["param"]` is a list so the order matter. Currently, the order is different. I suspect this is related to how NamedOptimizer creates the `param_groups`. Will debug this further when integrating with NamedOptimizer hooks.

fegin committed:
Update on "[FSDP][optim_state_dict][7/N] Make FSDP support NamedOptimizer"


**What does this PR do?**
This PR refactors FSDP optimizer state_dict APIs to accept `NamedOptimizer` as the input optimizer. The key difference is that the state_dict returned by `NamedOptimizer` is already keyed as FQN. This PR majorly changes the internal mapping to allows the optimizer state_dict to be keyed as FQN.


[ghstack-poisoned]

fegin committed:
Update on "[FSDP][optim_state_dict][7/N] Make FSDP support NamedOptimizer"


**What does this PR do?**
This PR refactors FSDP optimizer state_dict APIs to accept `NamedOptimizer` as the input optimizer. The key difference is that the state_dict returned by `NamedOptimizer` is already keyed as FQN. This PR majorly changes the internal mapping to allows the optimizer state_dict to be keyed as FQN.


[ghstack-poisoned]

fegin commented:
@pytorchbot merge
